navigation_and_routing/lib/src/app.dart

```diff
@@ -16,97 +28,238 @@ class Bookstore extends StatefulWidget {
-
-    _routerDelegate = SimpleRouterDelegate(
-      routeState: _routeState,
-      navigatorKey: _navigatorKey,
-      builder: (context) => BookstoreNavigator(
-        navigatorKey: _navigatorKey,
-      ),
-    );
-
-    // Listen for when the user logs out and display the signin screen.
-    _auth.addListener(_handleAuthStateChanged);
-
-    super.initState();
-  }
-
-  @override
-  Widget build(BuildContext context) => RouteStateScope(
-        notifier: _routeState,
-        child: BookstoreAuthScope(
-          notifier: _auth,
-          child: MaterialApp.router(
-            routerDelegate: _routerDelegate,
-            routeInformationParser: _routeParser,
-            // Revert back to pre-Flutter-2.5 transition behavior:
-            // https://github.com/flutter/flutter/issues/82053
-            theme: ThemeData(
-              pageTransitionsTheme: const PageTransitionsTheme(
-                builders: {
-                  TargetPlatform.android: FadeUpwardsPageTransitionsBuilder(),
-                  TargetPlatform.iOS: CupertinoPageTransitionsBuilder(),
-                  TargetPlatform.linux: FadeUpwardsPageTransitionsBuilder(),
-                  TargetPlatform.macOS: CupertinoPageTransitionsBuilder(),
-                  TargetPlatform.windows: FadeUpwardsPageTransitionsBuilder(),
+  Widget build(BuildContext context) {
+    return MaterialApp.router(
+      builder: (context, child) {
+        if (child == null) {
+          throw ('No child in .router constructor builder');
+        }
+        return BookstoreAuthScope(
+          notifier: auth,
+          child: child,
+        );
+      },
+      routerConfig: GoRouter(
+        refreshListenable: auth,
+        debugLogDiagnostics: true,
+        initialLocation: '/books/popular',
+        redirect: (context, state) {
+          final signedIn = BookstoreAuth.of(context).signedIn;
+          if (state.location != '/sign-in' && !signedIn) {
+            return '/sign-in';
+          }
+          return null;
+        },
+        routes: [
+          ShellRoute(
+            navigatorKey: appShellNavigatorKey,
+            builder: (context, state, child) {
+              return BookstoreScaffold(
+                child: child,
+              );
+            },
+            routes: [
+              ShellRoute(
+                pageBuilder: (context, state, child) {
+                  return FadeTransitionPage<dynamic>(
+                    key: state.pageKey,
+                    // Use a builder to get the correct BuildContext
+                    // TODO (johnpryan): remove when https://github.com/flutter/flutter/issues/108177 lands
+                    child: Builder(builder: (context) {
+                      return BooksScreen(
+                        child: child,
+                        onTap: (idx) {
+                          switch (idx) {
+                            case 1:
+                              GoRouter.of(context).go('/books/new');
```

Comment:
Suggestion: Would it be possible to use `context.go()` extension function instead of `GoRouter.of(context).go`? so it aligns with https://pub.dev/packages/go_router/example

Reply:
That fine with me